src/envoy/http/authn/http_filter.cc

```diff
@@ -48,49 +49,43 @@ FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
   filter_context_.reset(new Istio::AuthN::FilterContext(
       &headers, decoder_callbacks_->connection(), filter_config_));
 
-  authenticator_ = createPeerAuthenticator(
-      filter_context_.get(),
-      [this](bool success) { onPeerAuthenticationDone(success); });
-  authenticator_->run();
-
-  if (state_ == State::COMPLETE) {
-    return FilterHeadersStatus::Continue;
+  Payload payload;
+  authenticator_ = createPeerAuthenticator(filter_context_.get());
+  if (!authenticator_->run(&payload)) {
+    rejectRequest("Peer authentication failed.");
+    stopped_ = true;
+    return FilterHeadersStatus::StopIteration;
   }
 
-  stopped_ = true;
-  return FilterHeadersStatus::StopIteration;
-}
+  authenticator_ = createOriginAuthenticator(filter_context_.get());
+  bool success = authenticator_->run(&payload);
 
-void AuthenticationFilter::onPeerAuthenticationDone(bool success) {
-  ENVOY_LOG(debug, "{}: success = {}", __func__, success);
-  if (success) {
-    authenticator_ = createOriginAuthenticator(
-        filter_context_.get(),
-        [this](bool success) { onOriginAuthenticationDone(success); });
-    authenticator_->run();
-  } else {
-    rejectRequest("Peer authentication failed.");
-  }
-}
-void AuthenticationFilter::onOriginAuthenticationDone(bool success) {
-  ENVOY_LOG(debug, "{}: success = {}", __func__, success);
   // After Istio authn, the JWT headers consumed by Istio authn should be
   // removed.
   for (auto const iter : filter_config_.jwt_output_payload_locations()) {
```

Comment:
Can you add a TODO refer to https://github.com/istio/proxy/issues/1363 for this part. It need to be done even when peer authn fail. You can take care of it in this PR as well, if you want :)